zkvm/src/constraints.rs

```diff
@@ -89,21 +92,58 @@ impl Constraint {
         .map_err(|e| VMError::R1CSError(e))
     }
 
-    fn flatten<CS: r1cs::RandomizedConstraintSystem>(self, cs: &mut CS) -> r1cs::LinearCombination {
+    fn flatten<CS: r1cs::RandomizedConstraintSystem>(
+        self,
+        cs: &mut CS,
+    ) -> Result<(r1cs::LinearCombination, Option<Scalar>), r1cs::R1CSError> {
         match self {
-            Constraint::Eq(expr1, expr2) => expr1.to_r1cs_lc() - expr2.to_r1cs_lc(),
+            Constraint::Eq(expr1, expr2) => {
+                let assignment = expr1
+                    .eval()
+                    .and_then(|x| expr2.eval().map(|y| x.to_scalar() - y.to_scalar()));
+                Ok((expr1.to_r1cs_lc() - expr2.to_r1cs_lc(), assignment))
+            }
             Constraint::And(c1, c2) => {
-                let a = c1.flatten(cs);
-                let b = c2.flatten(cs);
+                let (a, a_assg) = c1.flatten(cs)?;
+                let (b, b_assg) = c2.flatten(cs)?;
                 let z = cs.challenge_scalar(b"ZkVM.verify.and-challenge");
-                a + z * b
+                let assignment = a_assg.and_then(|a| b_assg.map(|b| a + z * b));
+                Ok((a + z * b, assignment))
             }
             Constraint::Or(c1, c2) => {
-                let a = c1.flatten(cs);
-                let b = c2.flatten(cs);
+                let (a, a_assg) = c1.flatten(cs)?;
+                let (b, b_assg) = c2.flatten(cs)?;
                 // output expression: a * b
                 let (_l, _r, o) = cs.multiply(a, b);
-                r1cs::LinearCombination::from(o)
+                let assignment = a_assg.and_then(|a| b_assg.map(|b| a * b));
+                Ok((r1cs::LinearCombination::from(o), assignment))
+            }
+            Constraint::Not(c1) => {
+                let (x_lc, x_assg) = c1.flatten(cs)?;
+                let (xy_assg, xw_assg, y_assg) = match x_assg {
+                    Some(x_assg) => {
+                        let is_zero = x_assg.ct_eq(&Scalar::zero());
+                        let y_assg = Scalar::conditional_select(
+                            &Scalar::zero(),
+                            &Scalar::one(),
+                            is_zero.into(),
+                        );
+                        let w_assg =
+                            Scalar::conditional_select(&x_assg, &Scalar::one(), is_zero.into());
+                        let w_assg = w_assg.invert();
+                        (Some((x_assg, y_assg)), Some((x_assg, w_assg)), Some(y_assg))
+                    }
+                    None => (None, None, None),
+                };
+                let (l1, r1, o1) = cs.allocate_multiplier(xy_assg)?;
+                // constraint l1 to x
+                cs.constrain(l1 - x_lc);
+                // enforce `x*y == 0` (y is 0 if x != 0)
+                cs.constrain(o1.into());
+                let (_l2, _r2, o2) = cs.allocate_multiplier(xw_assg)?;
```

Comment:
maybe for clarity, allocate both multipliers first, then add 4 constraints and additionally comment that r2 is left unconstrained intentionally.